include/btllib/mi_bloom_filter.hpp

```diff
@@ -104,787 +76,600 @@ class MIBloomFilter
+  std::vector<size_t> get_id_occurence_count(const bool& include_saturated);
+
+  /** Returns an a filter size large enough to maintain an occupancy specified
    */
-  size_t get_id_counts_strand(std::vector<size_t>& counts) const
-  {
-    size_t saturated_counts = 0;
-    for (size_t i = 0; i < m_d_size; ++i) {
-      if (m_data[i] > MASK) {
-        ++counts[m_data[i] & ID_MASK];
-        ++saturated_counts;
-      } else {
-        ++counts[m_data[i] & ANTI_STRAND];
-      }
-    }
-    return saturated_counts;
-  }
+  static size_t calc_optimal_size(size_t entries,
+                                  unsigned hash_num,
+                                  double occupancy);
 
-  size_t get_pop() const
+private:
+  MIBloomFilter(const std::shared_ptr<MIBloomFilterInitializer>& mibfi);
+  static void save(const std::string& path,
+                   const cpptoml::table& table,
+                   const char* data,
+                   size_t n);
+  std::vector<uint64_t> get_rank_pos(const uint64_t* hashes) const;
+  uint64_t get_rank_pos(const uint64_t hash) const
   {
-    size_t index = m_bv.size() - 1;
-    while (m_bv[index] == 0) {
-      --index;
-    }
-    return m_rank_support(index) + 1;
+    return bv_rank_support(hash % il_bit_vector.size());
   }
+  std::vector<T> get_data(const std::vector<uint64_t>& rank_pos) const;
+  T get_data(const uint64_t& rank) const { return id_array[rank]; }
+  void set_data(const uint64_t& pos, const T& id);
+  void set_saturated(const uint64_t* hashes);
 
-  /*
-   * Mostly for debugging
-   * should equal get_pop if fully populated
-   */
-  size_t get_pop_non_zero() const
-  {
-    size_t count = 0;
-    for (size_t i = 0; i < m_d_size; ++i) {
-      if (m_data[i] != 0) {
-        ++count;
-      }
-    }
-    return count;
-  }
+  size_t id_array_size = 0;
+  size_t bv_size = 0;
+  unsigned kmer_size = 0;
+  unsigned hash_num = 0;
+  std::string hash_fn;
 
-  /*
-   * Checks data array for abnormal IDs
-   * (i.e. values greater than what is possible)
-   * Returns first abnormal ID or value of max_val if no abnormal IDs are found
-   * For debugging
-   */
-  T check_values(T max_val) const
-  {
-    for (size_t i = 0; i < m_d_size; ++i) {
-      if ((m_data[i] & ANTI_MASK) > max_val) {
-        return m_data[i];
-      }
-    }
-    return max_val;
-  }
+  sdsl::bit_vector bit_vector;
+  sdsl::bit_vector_il<BLOCKSIZE> il_bit_vector;
+  sdsl::rank_support_il<1> bv_rank_support;
+  std::unique_ptr<std::atomic<uint16_t>[]> counts_array;
+  std::unique_ptr<std::atomic<T>[]> id_array;
 
-  size_t get_pop_saturated() const
-  {
-    size_t count = 0;
-    for (size_t i = 0; i < m_d_size; ++i) {
-      if (m_data[i] > MASK) {
-        ++count;
-      }
-    }
-    return count;
-  }
+  bool bv_insertion_completed = false, id_insertion_completed = false;
+};
 
-  size_t size() const
-  {
-    return m_bv.size();
-  }
+bool
+MIBloomFilterInitializer::check_file_signature(
```

Comment:
At that time I had problems when having the implementations in the .cpp file. I cannot recall the reason for sure right now but counting Bloom filter also doesn't have a .cpp file probably because it also has some functions that must be implemented where they are declared.

Reply:
Oh ok, let's keep it in the headers then.